lighthouse-viewer/app/src/main.js

```diff
@@ -0,0 +1,518 @@
+    this.github = new GithubAPI();
+
+    this.addListeners();
+    this.loadFromURL();
+  }
+
+  addListeners() {
+    const printButton = document.querySelector('.js-print');
+    if (!printButton) {
+      return;
+    }
+
+    printButton.addEventListener('click', _ => {
+      window.print();
+    });
+
+    const button = document.createElement('button');
+    button.classList.add('share');
+    button.addEventListener('click', this.onShare);
+    printButton.parentElement.insertBefore(button, printButton);
+  }
+
+  loadFromURL() {
+    // Pull gist id from URL and render it.
+    const params = new URLSearchParams(location.search);
+    const gistId = params.get('gist');
+    if (gistId) {
+      logger.log('Fetching report from Github...', false);
+
+      this.github.auth.ready.then(_ => {
+        this.github.getGistContent(gistId).then(json => {
+          logger.hide();
+          this.replaceReportHTML(json.content);
+
+          // Save fetched json and etag to IDB so we can retrieve use it later
+          // for 304 requests.
+          return idb.set(gistId, {etag: json.etag, content: json.content});
+        }).catch(err => logger.error(err.message));
+      });
+    }
+  }
+
+  _validateReportJson(json) {
+    if (!json.lighthouseVersion) {
+      throw new Error('JSON file was not generated by Lighthouse');
+    }
+
+    // Leave off patch version in the comparison.
+    const semverRe = new RegExp(/^(\d+)?\.(\d+)?\.(\d+)$/);
+    const reportVersion = json.lighthouseVersion.replace(semverRe, '$1.$2');
+    const lhVersion = LH_CURRENT_VERSION.replace(semverRe, '$1.$2');
+
+    if (reportVersion < lhVersion) {
+      // TODO: figure out how to handler older reports. All permalinks to older
+      // reports will start to throw this warning when the viewer rev's its
+      // minor LH version.
+      // eslint-disable-next-line
+      window.alert('WARNING:  Results may not display properly.\n' +
+                  'Report was created with an earlier version of ' +
+                  `Lighthouse (${json.lighthouseVersion}). The latest ` +
+                  `version is ${LH_CURRENT_VERSION}.`);
+    }
+  }
+
+  replaceReportHTML(json) {
+    this._validateReportJson(json);
+
+    const reportGenerator = new ReportGenerator();
+
+    let html;
+    try {
+      html = reportGenerator.generateHTML(json, 'viewer');
+    } catch (err) {
+      html = reportGenerator.renderException(err, json);
+    }
+
+    // Use only the results section of the full HTML page.
+    const div = document.createElement('div');
+    div.innerHTML = html;
+    html = div.querySelector('.js-report').outerHTML;
+
+    this.json = json;
+
+    // Remove the placeholder drop area UI once the user has interacted.
+    this.fileUploader.removeDropzonePlaceholder();
+
+    // Replace the HTML and hook up event listeners to the new DOM.
+    document.querySelector('output').innerHTML = html;
+    this.addListeners();
```

Comment:
For the most part, all of the page is being replaced, including these UI buttons. It was much simple to nuke everything and re-attach than to replace parts of the page :)